redbpf/src/lib.rs

```diff
@@ -351,31 +356,37 @@ impl Program {
         if self.data().fd.is_some() {
             return Err(Error::ProgramAlreadyLoaded);
         }
+        // Should bind CString to local variable not to make a dangling pointer
+        // with .as_ptr() method
+        let cname = CString::new(self.data().name.clone())?;
         let clicense = CString::new(license)?;
-        let cname = CString::new(self.data_mut().name.clone())?;
-        let log_buffer: MutDataPtr =
-            unsafe { libc::malloc(mem::size_of::<i8>() * 16 * 65535) as MutDataPtr };
-        let buf_size = 64 * 65535_u32;
 
-        let fd = unsafe {
-            bpf_sys::bcc_prog_load(
-                self.to_prog_type(),
-                cname.as_ptr() as DataPtr,
-                self.data_mut().code.as_ptr(),
-                (self.data_mut().code.len() * mem::size_of::<bpf_insn>()) as i32,
-                clicense.as_ptr() as DataPtr,
-                kernel_version as u32,
-                0_i32,
-                log_buffer,
-                buf_size,
-            )
-        };
+        let mut attr = unsafe { mem::zeroed::<bpf_sys::bpf_load_program_attr>() };
 
-        if fd < 0 {
-            Err(Error::BPF)
-        } else {
-            self.data_mut().fd = Some(fd);
-            Ok(())
+        attr.prog_type = self.to_prog_type();
+        attr.expected_attach_type = 0;
+        attr.name = cname.as_ptr();
+        attr.insns = self.data().code.as_ptr();
+        attr.insns_cnt = self.data().code.len() as u64;
+        attr.license = clicense.as_ptr();
+        attr.__bindgen_anon_1.kern_version = kernel_version;
+        attr.log_level = 1;
+
+        unsafe {
+            let buf_size = 64 * 1024;
+            let log_buffer: MutDataPtr = libc::malloc(buf_size) as MutDataPtr;
+            let fd = bpf_sys::bpf_load_program_xattr(&attr, log_buffer, buf_size as u64);
+            if fd < 0 {
+                let cstr = CStr::from_ptr(log_buffer);
+                eprintln!("error loading BPF program {}", cstr.to_str().unwrap());
+                libc::free(log_buffer as *mut _);
+                Err(Error::BPF)
+            } else {
+                // free should be called to prevent memory leakage
+                libc::free(log_buffer as *mut _);
```

Comment:
Actually this is free-ing the log buffer of the program, which _may_ cause issues. I'd need to dig deeper to find the exact implications of this, but it might be worth keeping this buffer around in an internal struct somewhere. Have you checked if this is safe to do?

Reply:
I never considered that the output buffer kept being used by something inside libbpf after the `bpf_load_program_xattr` returns.
If it really is, the API is bug prone and there should be a lot of noticeable CAUTIONS around the function.

I'll check this is safe too. Thanks

Reply:
I checked that `log_buffer` is not used by libbpf. It is merely passed to `bpf` system call.

Inside linux kernel, `log_buf` of `bpf_attr` is used by BPF verifier and the buffer is not maintained by kernel after `bpf_check` is finished.

The call stack I checked is this
```
bpf_verifier_vlog (kernel/bpf/verifier.c)
bpf_check (kernel/bpf/verifier.c)
bpf_prog_load (kernel/bpf/syscall.c)
```

So I think it is fine to free `log_buffer` after `bpf_load_program_xattr` ends.